src/daft-local-execution/src/resource_manager.rs

```diff
@@ -0,0 +1,117 @@
+use std::sync::{Arc, Mutex, OnceLock};
+
+use common_error::{DaftError, DaftResult};
+use common_system_info::SystemInfo;
+use lazy_static::lazy_static;
+use tokio::sync::Notify;
+
+lazy_static! {
+    pub(crate) static ref MEMORY_MANAGER: OnceLock<Arc<MemoryManager>> = OnceLock::new();
+}
+
+fn custom_memory_limit() -> Option<u64> {
+    let memory_limit_var_name = "DAFT_MEMORY_LIMIT";
+    if let Ok(val) = std::env::var(memory_limit_var_name) {
+        if let Ok(val) = val.parse::<u64>() {
+            return Some(val);
+        }
+    }
+    None
+}
+
+pub(crate) fn get_or_init_memory_manager() -> &'static Arc<MemoryManager> {
+    MEMORY_MANAGER.get_or_init(|| Arc::new(MemoryManager::new()))
+}
+
+pub(crate) struct MemoryPermit<'a> {
+    bytes: u64,
+    manager: &'a MemoryManager,
+}
+
+impl Drop for MemoryPermit<'_> {
+    fn drop(&mut self) {
+        if self.bytes > 0 {
+            {
+                let mut state = self.manager.state.lock().unwrap();
+                state.available_bytes += self.bytes;
+            } // lock is released here
+            self.manager.notify.notify_waiters();
+        }
+    }
+}
+
+struct MemoryState {
+    available_bytes: u64,
+}
+
+pub(crate) struct MemoryManager {
```

Comment:
Let's write some rust tests here to test the MemoryManager's behavior and synchronization